src/units/attack_type.cpp

```diff
@@ -101,17 +101,16 @@ attack_type::attack_type(const config& cfg) :
 		else
 			icon_ = "attacks/blank-attack.png";
 	}
-	if(cfg.has_attribute("alignment") && (cfg["alignment"] == "neutral" || cfg["alignment"] == "lawful" || cfg["alignment"] == "chaotic" || cfg["alignment"] == "liminal")){
-		alignment_str_ = cfg["alignment"].str();
-	} else if(self_){
-		alignment_str_ =unit_alignments::get_string(self_->alignment());
-	}
 }
 
-unit_alignments::type attack_type::alignment() const
+utils::optional<unit_alignments::type> attack_type::alignment() const
 {
 	// pick attack alignment or fall back to unit alignment
-	return (unit_alignments::get_enum(alignment_str_).value_or(self_ ? self_->alignment() : unit_alignments::type::neutral));
+	utils::optional<unit_alignments::type> att_alignment = unit_alignments::get_enum(alignment_str_);
+	if(!att_alignment && self_){
```

Comment:
Since you made the callers decide on the fallback why still check `self_` here?

Reply:
Like unit alignment is class unit, it should check what self_ exist before to call this function.